src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderRepository.cs

```diff
@@ -0,0 +1,223 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using DBreeze;
+using DBreeze.DataTypes;
+using DBreeze.Utils;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders
+{
+    /// <summary>
+    /// Persistent implementation of the <see cref="ProvenBlockHeader"/> DBreeze repository.
+    /// </summary>
+    public class ProvenBlockHeaderRepository : IProvenBlockHeaderRepository
+    {
+        /// <summary>
+        /// Instance logger.
+        /// </summary>
+        private readonly ILogger logger;
+
+        /// <summary>
+        /// Access to DBreeze database.
+        /// </summary>
+        private readonly DBreezeEngine dbreeze;
+
+        /// <summary>
+        /// Specification of the network the node runs on - RegTest/TestNet/MainNet.
+        /// </summary>
+        private readonly Network network;
+
+        /// <summary>
+        /// Database key under which the block hash and height of a <see cref="ProvenBlockHeader"/> tip is stored.
+        /// </summary>
+        private static readonly byte[] blockHashHeightKey = new byte[0];
+
+        /// <summary>
+        /// DBreeze table names.
+        /// </summary>
+        private const string ProvenBlockHeaderTable = "ProvenBlockHeader";
+        private const string BlockHashHeightTable = "BlockHashHeight";
+
+        /// <summary>
+        /// Current <see cref="ProvenBlockHeader"/> tip.
+        /// </summary>
+        private ProvenBlockHeader provenBlockHeaderTip;
+
+        /// <inheritdoc />
+        public HashHeightPair TipHashHeight { get; private set; }
+
+        /// <summary>
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="network">Specification of the network the node runs on - RegTest/TestNet/MainNet.</param>
+        /// <param name="folder"><see cref="ProvenBlockHeaderRepository"/> folder path to the DBreeze database files.</param>
+        /// <param name="loggerFactory">Factory to create a logger for this type.</param>
+        public ProvenBlockHeaderRepository(Network network, string folder, ILoggerFactory loggerFactory)
+        {
+            Guard.NotNull(network, nameof(network));
+            Guard.NotNull(folder, nameof(folder));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            Directory.CreateDirectory(folder);
+
+            this.dbreeze = new DBreezeEngine(folder);
+            this.network = network;
+        }
+
+        /// <inheritdoc />
+        public Task InitializeAsync()
```

Comment:
This method will always initialize tip at 0 height. 

Please fix and add tests for this method. 
Test case: save 10 headers to repo, dispose the repo, create new repo, initialize it. Make sure repo's tip is at height 10.

Reply:
Ok, fixed.